crates/ursa-pod/src/codec.rs

```diff
@@ -1,2 +1,543 @@
+                pubkey,
+                compression_bitmap,
+                lane,
+            } => {
+                buf.reserve(55);
+                buf.put_slice(&NETWORK);
+                buf.put_u8(version);
+                buf.put_u8(compression_bitmap);
+                buf.put_u8(lane);
+                buf.put_slice(&pubkey);
+            }
+            UrsaFrame::HandshakeResponse {
+                pubkey,
+                epoch_nonce,
+                lane,
+                last,
+            } => {
+                let last = match last {
+                    None => [0x00].as_slice(),
+                    Some(_data) => [0x80].as_slice(),
+                };
+                buf.reserve(42 + last.len());
+                buf.put_u8(lane);
+                buf.put_u64(epoch_nonce);
+                buf.put_slice(&pubkey);
+                buf.put_slice(last);
+            }
+            UrsaFrame::ContentRequest { hash } => {
+                buf.put_slice(&hash);
+            }
+            UrsaFrame::ContentResponse {
+                proof,
+                compression,
+                signature,
+                content,
+            } => {
+                let proof = proof.to_bytes();
+                let proof_len = proof.len();
+                let content_len = content.len();
+                buf.reserve(67 + proof_len + content_len);
+                buf.put_u8(compression);
+                buf.put_slice(&signature);
+                buf.put_u64(proof_len as u64);
+                buf.put(proof);
+                buf.put_u64(content_len as u64);
+                buf.put_slice(&content);
+            }
+            UrsaFrame::RangeRequest {
+                hash,
+                chunk_start,
+                chunks,
+            } => {
+                buf.put_slice(&hash);
+                buf.put_u64(chunk_start);
+                buf.put_u16(chunks);
+            }
+            UrsaFrame::UpdateEpoch(nonce) => {
+                buf.put_u64(nonce);
+            }
+            UrsaFrame::DecryptionKeyRequest {
+                delivery_acknowledgement,
+            } => {
+                buf.put_slice(&delivery_acknowledgement);
+            }
+            UrsaFrame::DecryptionKeyResponse { decryption_key } => {
+                buf.put_slice(&decryption_key);
+            }
+            UrsaFrame::EndOfRequest => {}
+            UrsaFrame::Terminate(reason) => {
+                buf.put_u8(reason as u8);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl Decoder for UrsaCodec {
+    type Item = UrsaFrame;
+    type Error = UrsaCodecError;
+
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
+        let len = src.len();
+        if len == 0 {
+            return Ok(None);
+        }
+
+        // first frame byte is the tag
+        let tag = src[0];
+        match tag {
+            HANDSHAKE_REQ_TAG => {
+                if len < 56 {
```

Comment:
Could we please create consts for these magic numbers?

Reply:
We're not there yet, the proper plan will be to have a `impl Tag { pub fn frame_size(&self) -> usize {} }`. this is not a PR to main btw.

Reply:
also, the magic numbers are the byte totals documented on the `UrsaFrame` enum

Reply:
implemented FrameTag enum and size_hint method